cmake/Modules/PythonSetupSubdirectory.cmake

```diff
@@ -134,25 +134,31 @@ function(_generate_build_configuration_txt)
     # TODO(amp): Including INTERFACE_LINK_OPTIONS causes an unavoidable cmake generation failure in cmake 3.20+ due to the use of $<HOST_LINK:...>.
     #  INTERFACE_LINK_OPTIONS doesn't seem to be required for the dependencies we are using right now, so just skip it for the moment.
-    # list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_LINK_OPTIONS>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},LINKER:-rpath=$<TARGET_LINKER_FILE_DIR:${TARGET_NAME}>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},LINKER:-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},BUILD_RPATH>,;LINKER:-rpath=>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},LINKER:-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},INSTALL_RPATH>,;LINKER:-rpath=>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_LINKER_FILE:${TARGET_NAME}>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_LINK_DEPENDS>>")
-    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},LINK_DEPENDS>>")
-    list(APPEND COMPILE_OPTIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_COMPILE_OPTIONS>>")
+    # list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_LINK_OPTIONS>>")
+    if (NOT IS_IMPORTED)
+      # Imported libraries (pybind11 specifically) do not support these generator expressions.
+      # TODO(amp): This may prevent imported binary libraries from working in some corner cases. Who really knows with cmake.
+      list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},LINKER:-rpath=$<TARGET_LINKER_FILE_DIR:${TARGET_NAME}>>")
+      list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},LINKER:-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},BUILD_RPATH>,;LINKER:-rpath=>>")
+      list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},$<TARGET_LINKER_FILE:${TARGET_NAME}>>")
+    endif()
+    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},LINKER:-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},INSTALL_RPATH>,;LINKER:-rpath=>>")
+    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_LINK_DEPENDS>>")
+    list(APPEND LINK_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},LINK_DEPENDS>>")
+    # TODO(amp): Including INTERFACE_COMPILE_OPTIONS and using pybind11 causes LTO to be enabled, but only the compiler
+    #  option is actually passed to the setuptools compilers. So just disable this for the moment.
+    # list(APPEND COMPILE_OPTIONS "$<TARGET_GENEX_EVAL:${X_TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_COMPILE_OPTIONS>>")
```

Comment:
@ddn0 I do not like this hack, but I cannot for the life of me determine why the pybind11 LTO options are being picked up here. And turning it off doesn't seem to break things. Yet. What do you think?

Reply:
I'm not following. What is the issue with just propagating LTO here?

Reply:
Sorry. I didn't explain well.

LTO must be enabled at both compile and link time. If it's only enabled at compile time the linker will choke on the object files because LTO objects are totally different. We cannot pass through LINK_OPTIONS because some of the GPU related generator expressions (HOST_ONLY or something) cause an unavoidable CMake failure when used in custom files. So the linker does not get the LTO options set. Passing through the compiler options puts us in a state where the linker chokes, so I just didn't. I don't like it, but it works for now. Honestly the only fix I can see for all this is to stop using CMake or write a "dump flags" backend for CMake (in the codebase, there is no external backend support).